jitstats.lua

```diff
@@ -0,0 +1,462 @@
+  {name = "MCODEAL", msg = "failed to allocate mcode memory"},
+  {name = "MCODEOV", msg = "machine code too long"},
+  {name = "MCODELM", msg = "hit mcode limit (retrying)"},
+  {name = "SPILLOV", msg = "too many spill slots"},
+  {name = "BADRA",   msg = "inconsistent register allocation"},
+  {name = "NYIIR",   msg = "NYI: cannot assemble IR instruction %d"},
+  {name = "NYIPHI",  msg = "NYI: PHI shuffling too complex"},
+  {name = "NYICOAL", msg = "NYI: register coalescing too complex"},
+}
+
+local traceerr, traceerr_fromid = {}, table.new(#vmdef.traceerr, 0)
+
+for i, evt in pairs(traceerror_info) do
+    local id = traceerr_lookup[evt.msg]
+
+    if id then
+        traceerr[evt.name] = id
+        traceerr_fromid[id] = evt.name
+    end
+end
+
+local TRERR_BLACKL, TRERR_LLEAVE, TRERR_LUNROLL, TRERR_NYIBC = traceerr.BLACKL, traceerr.LLEAVE, traceerr.LUNROLL, traceerr.NYIBC
+
+local function printf(fmt, ...)
+    if select("#", ...) == 0 then
+        print(fmt)
+    else
+        print(string.format(fmt, ...))
+    end
+end
+
+local debugprint = function () end
+
+-- Stolen from dump.lua
+local function fmtfunc(func, pc)
+    local fi = funcinfo(func, pc)
+    if fi.loc then
+        return fi.loc
+    elseif fi.ffid then
+        return vmdef.ffnames[fi.ffid]
+    elseif fi.addr then
+        return ("C:%x"):format(fi.addr)
+    else
+        return "(?)"
+    end
+end
+
+-- Format trace error message.  Stolen from dump.lua
+local function fmterr(err, info)
+    if type(err) == "number" then
+        if type(info) == "function" then 
+            info = fmtfunc(info) 
+        end
+        err = vmdef.traceerr[err]:format(info)
+    end
+    return err
+end
+
+local jitstats = {
+    isrunning = false,
+    debug = false,
+    stats = {}
+}
+
+local stats = jitstats.stats
+
+local vmevent = {}
+
+local curr_parentid, curr_stitched
+local start_pc, start_func
+-- Trace started
+function vmevent.start(tr, func, pc, parentid, exitnum)
+    curr_parentid = parentid
+    start_func = func
+    start_pc = pc
+    curr_stitched = parentid and parentid > 0 and exitnum == -1
+
+    debugprint("START(%d): parent = %d", tr, parentid or -1)
+    
+    stats.starts = stats.starts + 1
+    if curr_stitched then
+        stats.stitch_starts = stats.stitch_starts + 1
+    elseif parentid then
+        stats.side_starts = stats.side_starts + 1
+    end 
+end
+
+--Trace Stopped
+function vmevent.stop(tr)
+    local info = jutil.traceinfo(tr)
+     
+    debugprint("STOP(%d): link = %s", tr, info.linktype)
+    
+    stats.completed = stats.completed + 1
+    
+    if curr_stitched then
+        stats.stitch_completed = stats.stitch_completed + 1
+    elseif curr_parentid then
+        stats.side_completed= stats.side_completed + 1
+    end
+
+    stats.linktypes[info.linktype] = (stats.linktypes[info.linktype] or 0) + 1
+end
+
+--Trace Aborted
+function vmevent.abort(tr, func, pc, code, errinfo)
+    stats.aborts = stats.aborts + 1
+    
+    local reason
```

Comment:
don't leave uninitialized.  it doesn't matter right now, but a change in the `debugprint()` behaviour could fail.  IOW: non-local invariant violation.

Reply:
fixed